server/jobs/moss.py

```diff
@@ -0,0 +1,123 @@
+
+    seen = set()
+    for subm in subms:
+        if subm['backup']['id'] in seen:
+            continue
+        else:
+            seen.add(subm['backup']['id'])
+
+        if subm['group']:
+            logger.info("{} -> {}".format(encode_id(subm['backup']['id']),
+                                          subm['group']['group_member_emails']))
+        else:
+            logger.info("{} -> {}".format(encode_id(subm['backup']['id']),
+                                          subm['user']['email']))
+
+    backup_query = (Backup.query.options(db.joinedload('messages'))
+                          .filter(Backup.id.in_(subm_keys))
+                          .order_by(Backup.created.desc())
+                          .all())
+
+    logger.info("Retreived {} final submissions".format(len(subm_keys)))
+
+    # TODO: Customize the location of the tmp writing (especially useful during dev)
+
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        # Copy in the moss script
+        with open('server/jobs/moss-submission.pl', 'r') as f:
+            moss_script = f.read()
+            moss_script = moss_script.replace('YOUR_USER_ID_HERE', str(moss_id))
+            with open(tmp_dir + "/moss.pl", 'w') as script:
+                script.write(moss_script)
+
+        for backup in backup_query:
+            # Write file into file
+            file_contents = [m for m in backup.messages if m.kind == 'file_contents']
+            if not file_contents:
+                logger.info("{} didn't have any file contents".format(backup.hashid))
+                continue
+            contents = file_contents[0].contents
+            dest_dir = "{}/{}/".format(tmp_dir, backup.hashid)
+
+            if not os.path.exists(dest_dir):
+                os.makedirs(dest_dir)
+
+            for file in contents:
+                if file == 'submit':  # ignore fake file from ok-client
+                    continue
+                with open(dest_dir + file, 'w') as f:
+                    f.write(contents[file])
+
+        logger.info("Wrote all files to {}".format(tmp_dir))
+
+        template_files = []
+        for template in assign.files:
+            dest = "{}/{}".format(tmp_dir, template)
+            with open(dest, 'w') as f:
+                f.write(assign.files[template])
+
+            template_files.append(template)
+
+        logger.info("Using template files: {}".format(' '.join(template_files)))
+
+        templates = ' '.join(["-b {file}".format(file=f) for f in template_files])
+
+        os.chdir(tmp_dir)
+        all_student_files = glob.glob("*/*")
+        chosen_files, available_file_names = [], []
+        match_pattern = re.compile(file_regex)
+
+        logger.info("Running with regex match {}".format(file_regex))
+
+        for file in all_student_files:
+            fname = ntpath.basename(file)
+            available_file_names.append(fname)
+            if match_pattern.match(fname):
```

Comment:
This match should be done while copying the backup files over, so that we don't need to copy files we aren't going to use. I think it'll make the logic easier to follow too (no need to differentiate between `all_student_files` and `chosen_files`)